File: leetcode/Merge_Intervals.java
import java.util.*;

/**
 * Definition for an interval.
 * public class Interval {
 *     int start;
 *     int end;
 *     Interval() { start = 0; end = 0; }
 *     Interval(int s, int e) { start = s; end = e; }
 * }
 */

/*
class Interval {
    int start;
    int end;
    Interval() { start = 0; end = 0; }
    Interval(int s, int e) { start = s; end = e; }
}*/

class IntervalComparator implements Comparator<Interval> {
    public int compare(Interval x1, Interval x2) {
        if (x1.start == x2.start) return x2.end - x1.end;
        return x1.start - x2.start;
    }
}

public class Solution {
    public ArrayList<Interval> merge(ArrayList<Interval> intervals) {
        ArrayList<Interval> ret = new ArrayList<Interval>();
        //if (intervals.size() > 0) return ret;
        Collections.sort(intervals, new IntervalComparator());
        ListIterator<Interval> it = intervals.listIterator();
        Interval cur = null;
        while (it.hasNext()) {
            Interval in = it.next();
            if (cur == null) {
                cur = new Interval(in.start, in.end);
            } else {
                if (in.start <= cur.end) {
                    cur.end = Math.max(cur.end, in.end);
                } else {
                    ret.add(cur);
                    cur = new Interval(in.start, in.end);
                }
            }
        }
        if (cur != null) ret.add(cur);
        return ret;
    }
}

File: leetcode/Merge_Intervals.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */

bool compareIntervals(const Interval &a, const Interval &b) {
    if (a.start == b.start)
        return a.end > b.end;
    return a.start < b.start;
}

class Solution {
public:
    vector<Interval> merge(vector<Interval> &intervals) {
        int n = intervals.size();
        if (n <= 1)
            return intervals;
        sort(intervals.begin(), intervals.end(), compareIntervals);
        vector<Interval> after_mergence;
        after_mergence.push_back(intervals[0]);
        for (int i = 1; i < n; ++i)
            if (intervals[i].start > after_mergence.back().end)
                after_mergence.push_back(intervals[i]);
            else
                after_mergence.back().end = max(after_mergence.back().end, 
                                                intervals[i].end);
        return after_mergence;
    }
};
